### PR TITLE
Ataxia

### PR DESCRIPTION
**Pull request by ihaywood3**
This is Ataxia, a proposed alternative/replacement for Staggered Orders. (happy to change the name)
See my brief "whitepaper" [ataxia.pdf](https://github.com/Codaone/DEXBot/files/2152786/ataxia.pdf)
this code is slightly different: the market last price is used for the centre price, unless this isn't available, in which case it uses the midpoint of the range, so it can do true bootstrapping (i.e form orders on empty markets)

### CODE REVIEW

**Comment by MarkoPaasila** [comment]
@ihaywood3 do you think it would be ok to leave Ataxia without a GUI until we have UI autogeneration? Just cli.

File: dexbot/strategies/ataxia.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
+import math
+import time
+
+from bitshares.amount import Amount
```
Comment (mikakoi) [inline dexbot/strategies/ataxia.py]: Unused

File: dexbot/strategies/ataxia.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+                return True
+            if self[old_param] != self.worker[param]:
+                self.save_params()
+                return True
+        return False
+
+    def check_at_price(self, price):
+        """True if no order in self.orderlist at this price"""
+        for o in self.orders:
+            if abs(o['price']-price)/price < 0.001:  # "within 0.1% means equal" as slight errors creep in due to rounding
```
Comment (mikakoi) [inline dexbot/strategies/ataxia.py]: The line exceeds 120 characters

File: dexbot/strategies/ataxia.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        # don't ever trade against the market
+        if 'highestBid' in ticker and ticker['highestBid']:
+            bid = float(ticker['highestBid'])
+            if bid > 0.0:
+                lowest_sell = max(lowest_sell, bid)
+        if 'lowestAsk' in ticker and ticker['lowestAsk']:
+            ask = float(ticker['lowestAsk'])
+            if ask > 0.0:
+                highest_buy = min(highest_buy, ask)
+        return (highest_buy, lowest_sell)
```
Comment (mikakoi) [inline dexbot/strategies/ataxia.py]: Redundant parentheses

File: dexbot/strategies/ataxia.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+    def compute_bounds(self):
+        """
+        In descendants, allow the bounds to be dynamic by some method
+        (maybe external price feed, maybe some other market analysis)
+        In this implementation, just return the values unmodified
+        """
+        return (self.upper_bound, self.lower_bound)
```
Comment (mikakoi) [inline dexbot/strategies/ataxia.py]: Redundant parentheses

File: dexbot/strategies/ataxia.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+            new_order = False
+            self.account.refresh()
+            highest_buy, lowest_sell = self.spread_zone(upper_bound, lower_bound)
+            self.log.debug("highest_buy = {} lowest_sell = {}".format(highest_buy, lowest_sell))
+            # do max one order on each side, then cycle outer loop (i.e. check back
+            # with market whether things have shifted)
+            for price, ladder_size in downladder:
+                if price > lowest_sell:
+                    if self.check_at_price(1/price):  # sell orders are inverted
+                        total_amount = size*ladder_size  # ladder sizes from base of 1, multiple by size toget "real" order size
```
Comment (mikakoi) [inline dexbot/strategies/ataxia.py]: The line exceeds 120 characters

File: dexbot/strategies/ataxia.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+        """ Handle market update callbacks
+        """
+        delta = time.time() - self.last_check
+
+        # Only allow to check orders whether minimal time passed
+        if delta > CHECK_MIN_TIME:
+            self.reassess()
+
+    @classmethod
+    def get_required_assets(cls, *args, **kwargs):
```
Comment (mikakoi) [inline dexbot/strategies/ataxia.py]: What's the point of this method?

File: dexbot/strategies/ataxia.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+        self.size = self.worker['size']
+        self.spread = self.worker['spread'] / 100
+        self.increment = self.worker['increment'] / 100
+        self.upper_bound = self.worker['upper_bound']
+        self.lower_bound = self.worker['lower_bound']
+        # Order expiration time, should be high enough
+        self.expiration = 60*60*24*365*5
+        self.last_check = 0
+
+        if kwargs.get('active', True):
```
Comment (mikakoi) [inline dexbot/strategies/ataxia.py]: This isn't defined anywhere?

**Comment by mikakoi** [comment]
@ihaywood3 I made some comments but this also requires changes to the configure method. I merged the UI auto generator code to master which changes the ConfigElement structure. If you're able to make the needed changes I'll be able to merge this.

**Comment by ihaywood3** [comment]
In Staggered Orders a static method is used to compute required assets
during configuration
I found this got too complex for ataxia and IMHO it was better to
instantiate as an 'inactive' non-trading strategy
object, then the same routines cna be used to compute the required assets
for the given settings,
this is not finished and happy to remove it but it's an aide memoire to use
this approach when/if this is required,

On Fri, Jul 27, 2018 at 9:36 PM, Mika Koivistoinen <notifications@github.com

**Comment by ihaywood3** [comment]
respectfully I disagree I find readability suffers without parentheses.

OK I'm betraying my LISP/Prolog influences: I just like parentheses

On Fri, Jul 27, 2018 at 9:05 PM, Mika Koivistoinen <notifications@github.com

**Comment by ihaywood3** [comment]
ACK
will let you know when I have updated the config metadata and fixed what
autopep8 missed.

On Fri, Jul 27, 2018 at 9:47 PM, Mika Koivistoinen <notifications@github.com

**Comment by mikakoi** [comment]
You're free to use the parentheses if you want to, I just pointed out that they're redundant.

**Comment by mikakoi** [comment]
@ihaywood3 Did you just accidentally the whole repo?

**Comment by ihaywood3** [comment]
I tried to merge changes but caused more problems than I solved.
I have used git reset --hard to hopefully undo all this and just uploaded
my latest version of ataxia

On Thu, Aug 2, 2018 at 10:31 PM, Mika Koivistoinen <notifications@github.com

**Comment by mikakoi** [comment]
You added bunch of changes that aren't related to ataxia, could you remove those changes?

**Pull request closed without merging**